docs/library/machine.I2S.rst

```diff
@@ -0,0 +1,157 @@
+
+I2S is a synchronous serial protocol used to connect digital audio devices. 
+At the physical level, a bus consists of 3 lines: SCK, WS, SD.
+The I2S class supports Master operation.  Slave operation is not supported.
+
+I2S objects can be created and initialized using::
+
+    from machine import I2S
+    from machine import Pin
+    
+    # ESP32
+    sck_pin = Pin(14)   # Serial clock output
+    ws_pin = Pin(13)    # Word clock output
+    sdout_pin = Pin(12) # Serial data output
+    
+    or
+    
+    # PyBoards
+    sck_pin = Pin("Y6")   # Serial clock output
+    ws_pin = Pin("Y5")    # Word clock output
+    sdout_pin = Pin("Y8") # Serial data output
+    
+    audio_out = I2S(2, 
+                    sck=sck_pin, ws=ws_pin, sdin=sdin_pin,    
+                    mode=I2S.TX, 
+                    bits=16,                       
+                    format=I2S.MONO,
+                    rate=44100, 
+                    bufferlen=20000)
+                   
+    audio_in = I2S(2, 
+                   sck=sck_pin, ws=ws_pin, sdin=sdin_pin,    
+                   mode=I2S.RX, 
+                   bits=32,                       
+                   format=I2S.STEREO,
+                   rate=22050, 
+                   bufferlen=20000)
+                    
+3 modes of operation are supported:
+ - blocking 
+ - non-blocking 
+ - uasyncio
+  
+blocking::
+ 
+   num_written = audio_out.write(buf) # blocks until buf emptied
+
+   num_read = audio_in.readinto(buf) # blocks until buf filled
+   
+non-blocking::
+ 
+   audio_out.irq(i2s_callback)         # i2s_callback is called when buf is emptied
+   num_written = audio_out.write(buf)  # returns immediately
+        
+   audio_in.irq(i2s_callback)          # i2s_callback is called when buf is filled
+   num_read = audio_in.readinto(buf)   # returns immediately    
+ 
+uasyncio::
+ 
+   swriter = uasyncio.StreamWriter(audio_out)
+   swriter.write(buf)
+   await swriter.drain()
+   
+   sreader = uasyncio.StreamReader(audio_in)
+   num_read = await sreader.readinto(buf)
+ 
+Constructor
+-----------
+
+.. class:: I2S(id, *, sck, ws, sd, mode, bits, format, rate, bufferlen)
+
+   Construct an I2S object of the given id:
+   
+   - ``id`` identifies a particular I2S bus.  
+
+   ``id`` is board and port specific:
+ 
+     - PYBv1.0/v1.1: has one I2S bus with id=2.
+     - PYBD-SFxW: has two I2S busses with id=1 and id=2. 
+     - ESP32: has two I2S busses with id=0 and id=1. 
+   
+   Keyword-only parameters that are supported on all ports:
+    
+     - ``sck`` is a pin object for the serial clock line
+     - ``ws`` is a pin object for the word select line
+     - ``sd`` is a pin object for the serial data line
+     - ``mode`` specifies receive or transmit
+     - ``bits`` specifies sample size (bits), 16 or 32
+     - ``format`` specifies channel format, STEREO or MONO
+     - ``rate`` specifies audio sampling rate (samples/s)
+     - ``bufferlen`` specifies internal buffer length (bytes)
```

Comment:
What do you think about renaming this `bufferlen` argument to `buffer` or just `buf`?  In `machine.UART` there's a similar argument `rxbuf` which sets the incoming buffer size for the UART.  The reason for calling it `rxbuf` instead of `rxbuf_len` is that the former is shorter, and also it allows the possibility to support passing in an actual pre-allocated buffer rather than in integer.

So for I2S it could be:
```python
i2s = machine.I2S(..., buf=1024)  # ask for buffer to be allocated to given size
i2s = machine.I2S(..., buf=bytearray(1024))  # provide a buffer ourselves
```
And then the user could use some additional RAM that's not part of the uPy GC heap (eg at 0x10000000) for a (large) buffer.

Reply:
> What do you think about renaming this `bufferlen` argument to `buffer` or just `buf`

That works well.  I decided to go with `ibuf` (internal buffer) to reduce confusion with the sample buffers used in the `write()` and `readinto()` methods.

Reply:
Sounds good!